// nuggies/src/classes/buttonroles.js
const { MessageButton, MessageActionRow } = require('discord-buttons');
const { MessageEmbed, Message, Channel } = require('discord.js');

const Discord = require("discord.js");
const client = new Discord.Client();
const get = require('axios');
const mysql = require('mysql');
var req = require("request");
const variables = require(`../../../../variables/variables.json`);
const emoji = require(`../../../../variables/emoji.json`);
const Role = require(`../../../../variables/role.json`);
const Trad = require(`../../../../language/fr/buttonrole.json`);
const API = variables.API; // API FREE  (REGLAGE DE L'API AUTH.GG POUR GERER L'APP DEPUIS DISCORD) :)

class buttonroles {

	constructor() {
		this.roles = [];
		return this;
	}

	/**
	 *
	 * @param {String} color - Button Color [optional]
	 * @param {String} label - Button label
	 * @param {String} emoji - The emoji id [optional]
	 * @param {String} role - The role id
	 */
	addrole({ color, label, emoji, role }) {
		if (!color) color = 'gray';
		if (!label) throw new Error('please provide the button label!');
		if (!emoji) emoji = null;
		if (!role) throw new Error('please provide a role!');
		this.roles.push({ color: color, label: label, emoji: emoji, role: role });
		return this;
	}
	toJSON() { return { roles: this.roles }; }

	/**
	 * @param {Message} message - The Discord Message
	 * @param {MessageEmbed} embed - The Discord Embed
	 * @param {buttonroles} role - The created object using .buttonroles().addrole()
	 * @param {String} channelID - the id of the channel you want to send the message to.
	 */
	static async create({ message, content, role, channelID }) {
		if((message instanceof Message) == false) throw new TypeError('please provide the Discord Message');
		if(!content) throw new Error('please provide content!');
		if(!role) throw new Error('role not provided!');
		if(!channelID) throw new Error('channelID not provided!');
		const buttons = [];
		const rows = [];
		// Promise.resolve(role).then(console.log);
		// console.log(role);
		for (const buttonObject of role.roles) {
			buttons.push(new MessageButton().setStyle(buttonObject.color).setEmoji(buttonObject.emoji).setLabel(buttonObject.label).setID(`br:${buttonObject.role}`));
		}
		for (let i = 0; i < Math.ceil(role.roles.length / 5); i++) {
			rows.push(new MessageActionRow());
		}
		rows.forEach((row, i) => {
			row.addComponents(buttons.slice(0 + (i * 5), 5 + (i * 5)));
		});
		content instanceof MessageEmbed ? message.client.channels.cache.get(channelID).send({ embed: content, components: rows }) : message.client.channels.cache.get(channelID).send(content, { components: rows });
	}
	static async buttonclick(client, button, message) {
		if(!client) throw new Error('client not provided in buttonclick!');
		if(!button) throw new Error('Button not provided!');
		await button.clicker.fetch();
		const id = button.id;
		if (id.startsWith('br')) {
			let member;
			const fetchMem = await button.guild.members.fetch(button.clicker.member.id, false);
			if (fetchMem) member = button.guild.members.cache.get(button.clicker.member.id);
			await member.fetch(true);
			const role = id.split(':')[1];
			
			
				

				//===========================
    
				// Lien API

				get(`https://developers.auth.gg/USERS/?type=fetch&authorization=${API}&user=${button.clicker.member.id}`, {
					headers: {
						'Content-Type': "application/json",
					}
				}).then( (res) => {
					//-----------------------------//
					//           STATS             //
					//-----------------------------//
					try {
			
						//console.log(res.data)
						if(button.clicker.member.roles.cache.has(Role.IDAdmin))
						{
							if(variables.AdminAccesAllInfos === "yes")
							{
								button.reply.send(`${Trad.text33}`, true);
							}
						}
						else
						//+++
						if(res.data.rank === "1")
						{
							if(button.clicker.member.roles.cache.has(Role.IDBlanchisseur))
							{
								button.reply.send(`${Trad.text34} <@&${Role.IDBlanchisseur}>.`, true);
							}
							else
							{
								button.clicker.member.roles.add(Role.IDBlanchisseur);
								button.reply.send(`${Trad.text35} <@&${Role.IDBlanchisseur}>.`, true);
								if(button.clicker.member.roles.cache.has(Role.IDBlanchisseur))
								{
									//Si il a deja le role ont fait aucune action
								}
								else
								{
									button.clicker.member.roles.add(Role.IDPriver);
								}
							}
							
						}
						else
						if(res.data.rank === "2")
						{
							if(button.clicker.member.roles.cache.has(Role.IDAllWeed))
							{
								button.reply.send(`${Trad.text34} <@&${Role.IDAllWeed}>.`, true);
							}
							else
							{
								button.clicker.member.roles.add(Role.IDAllWeed);
								button.reply.send(`${Trad.text35} <@&${Role.IDAllWeed}>.`, true);
								if(button.clicker.member.roles.cache.has(Role.IDBlanchisseur))
								{
									//Si il a deja le role ont fait aucune action
								}
								else
								{
									button.clicker.member.roles.add(Role.IDPriver);
								}
							}
							
						}
						else
						if(res.data.rank === "3")
						{
							if(button.clicker.member.roles.cache.has(Role.IDWeed1))
							{
								button.reply.send(`${Trad.text34} <@&${Role.IDWeed1}>.`, true);
							}
							else
							{
								button.clicker.member.roles.add(Role.IDWeed1);
								button.reply.send(`${Trad.text35} <@&${Role.IDWeed1}>.`, true);
								if(button.clicker.member.roles.cache.has(Role.IDBlanchisseur))
								{
									//Si il a deja le role ont fait aucune action
								}
								else
								{
									button.clicker.member.roles.add(Role.IDPriver);
								}
							}
							
						}
						else
						if(res.data.rank === "4")
						{
							if(button.clicker.member.roles.cache.has(Role.IDWeed2))
							{
								button.reply.send(`${Trad.text34} <@&${Role.IDWeed2}>.`, true);
							}
							else
							{
								button.clicker.member.roles.add(Role.IDWeed2);
								button.reply.send(`${Trad.text35} <@&${Role.IDWeed2}>.`, true);
								if(button.clicker.member.roles.cache.has(Role.IDBlanchisseur))
								{
									//Si il a deja le role ont fait aucune action
								}
								else
								{
									button.clicker.member.roles.add(Role.IDPriver);
								}
							}
							
						}
						else
						if(res.data.rank === "5")
						{
							if(button.clicker.member.roles.cache.has(Role.IDWeed3))
							{
								button.reply.send(`${Trad.text34} <@&${Role.IDWeed3}>.`, true);
							}
							else
							{
								button.clicker.member.roles.add(Role.IDWeed3);
								button.reply.send(`${Trad.text35} <@&${Role.IDWeed3}>.`, true);
								if(button.clicker.member.roles.cache.has(Role.IDBlanchisseur))
								{
									//Si il a deja le role ont fait aucune action
								}
								else
								{
									button.clicker.member.roles.add(Role.IDPriver);
								}
							}
							
						}
						else
						if(res.data.rank === "6")
						{
							if(button.clicker.member.roles.cache.has(Role.IDAllOpiume))
							{
								button.reply.send(`${Trad.text34} <@&${Role.IDAllOpiume}>.`, true);
							}
							else
							{
								button.clicker.member.roles.add(Role.IDAllOpiume);
								button.reply.send(`${Trad.text35} <@&${Role.IDAllOpiume}>.`, true);
								if(button.clicker.member.roles.cache.has(Role.IDBlanchisseur))
								{
									//Si il a deja le role ont fait aucune action
								}
								else
								{
									button.clicker.member.roles.add(Role.IDPriver);
								}
							}
							
						}
						else
						if(res.data.rank === "7")
						{
							if(button.clicker.member.roles.cache.has(Role.IDOpiume1))
							{
								button.reply.send(`${Trad.text34} <@&${Role.IDOpiume1}>.`, true);
							}
							else
							{
								button.clicker.member.roles.add(Role.IDOpiume1);
								button.reply.send(`${Trad.text35} <@&${Role.IDOpiume1}>.`, true);
								if(button.clicker.member.roles.cache.has(Role.IDBlanchisseur))
								{
									//Si il a deja le role ont fait aucune action
								}
								else
								{
									button.clicker.member.roles.add(Role.IDPriver);
								}
							}
							
						}
						else
						if(res.data.rank === "8")
						{
							if(button.clicker.member.roles.cache.has(Role.IDOpiume2))
							{
								button.reply.send(`${Trad.text34} <@&${Role.IDOpiume2}>.`, true);
							}
							else
							{
								button.clicker.member.roles.add(Role.IDOpiume2);
								button.reply.send(`${Trad.text35} <@&${Role.IDOpiume2}>.`, true);
								if(button.clicker.member.roles.cache.has(Role.IDBlanchisseur))
								{
									//Si il a deja le role ont fait aucune action
								}
								else
								{
									button.clicker.member.roles.add(Role.IDPriver);
								}
							}
							
						}
						else
						if(res.data.rank === "9")
						{
							if(button.clicker.member.roles.cache.has(Role.IDOpiume3))
							{
								button.reply.send(`${Trad.text34}<@&${Role.IDOpiume3}>.`, true);
							}
							else
							{
								button.clicker.member.roles.add(Role.IDOpiume3);
								button.reply.send(`${Trad.text35} <@&${Role.IDOpiume3}>.`, true);
								if(button.clicker.member.roles.cache.has(Role.IDBlanchisseur))
								{
									//Si il a deja le role ont fait aucune action
								}
								else
								{
									button.clicker.member.roles.add(Role.IDPriver);
								}
							}
							
						}
						else
						if(res.data.rank === "10")
						{
							if(button.clicker.member.roles.cache.has(Role.IDAllEcstasy))
							{
								button.reply.send(`${Trad.text34} <@&${Role.IDAllEcstasy}>.`, true);
							}
							else
							{
								button.clicker.member.roles.add(Role.IDAllEcstasy);
								button.reply.send(`${Trad.text35} <@&${Role.IDAllEcstasy}>.`, true);
								if(button.clicker.member.roles.cache.has(Role.IDBlanchisseur))
								{
									//Si il a deja le role ont fait aucune action
								}
								else
								{
									button.clicker.member.roles.add(Role.IDPriver);
								}
							}
							
						}
						else
						if(res.data.rank === "11")
						{
							if(button.clicker.member.roles.cache.has(Role.IDEcstasy1))
							{
								button.reply.send(`${Trad.text34} <@&${Role.IDEcstasy1}>.`, true);
							}
							else
							{
								button.clicker.member.roles.add(Role.IDEcstasy1);
								button.reply.send(`${Trad.text35} <@&${Role.IDEcstasy1}>.`, true);
								if(button.clicker.member.roles.cache.has(Role.IDBlanchisseur))
								{
									//Si il a deja le role ont fait aucune action
								}
								else
								{
									button.clicker.member.roles.add(Role.IDPriver);
								}
							}
							
						}
						else
						if(res.data.rank === "12")
						{
							if(button.clicker.member.roles.cache.has(Role.IDEcstasy2))
							{
								button.reply.send(`${Trad.text34} <@&${Role.IDEcstasy2}>.`, true);
							}
							else
							{
								button.clicker.member.roles.add(Role.IDEcstasy2);
								button.reply.send(`${Trad.text35} <@&${Role.IDEcstasy2}>.`, true);
								if(button.clicker.member.roles.cache.has(Role.IDBlanchisseur))
								{
									//Si il a deja le role ont fait aucune action
								}
								else
								{
									button.clicker.member.roles.add(Role.IDPriver);
								}
							}
							
						}
						else
						if(res.data.rank === "13")
						{
							if(button.clicker.member.roles.cache.has(Role.IDEcstasy3))
							{
								button.reply.send(`${Trad.text34} <@&${Role.IDEcstasy3}>.`, true);
							}
							else
							{
								button.clicker.member.roles.add(Role.IDEcstasy3);
								button.reply.send(`${Trad.text35} <@&${Role.IDEcstasy3}>.`, true);
								if(button.clicker.member.roles.cache.has(Role.IDBlanchisseur))
								{
									//Si il a deja le role ont fait aucune action
								}
								else
								{
									button.clicker.member.roles.add(Role.IDPriver);
								}
							}
							
						}
						else
						if(res.data.rank === "14")
						{
							button.reply.send(`${Trad.text36}`, true);
						}
						else
						if(res.data.rank === "15")
						{
							button.reply.send(`${Trad.text36}`, true);
						}
						else
						if(res.data.rank === "16")
						{
							button.reply.send(`${Trad.text36}`, true);
						}
						else
						if(res.data.rank === "17")
						{
							button.reply.send(`${Trad.text36}`, true);
						}
						else
						if(res.data.rank === "18")
						{
							button.reply.send(`${Trad.text36}`, true);
						}
						else
						if(res.data.rank === "19")
						{
							button.reply.send(`${Trad.text36}`, true);
						}
						else
						if(res.data.rank === "20")
						{
							button.reply.send(`${Trad.text36}`, true);
						}
						else
						if(res.data.info === "No user found")
						{
							//+++
							if (button.clicker.member.roles.cache.has(role)) {
								button.reply.send(`${Trad.text37} ${Trad.text34} <@&${role}>.`, true);
							}
							else
							{
								button.clicker.member.roles.add(role);
								button.reply.send(`${Trad.text35} <@&${role}>.`, true);
							}
							//+++
						}
						else
						//+++
						{
							button.reply.send(`${Trad.text38} ${emoji.no}`, true);
						}
					//+++
					}
				//-----
				//--Fin du TRY
				//-----
					
					catch(error)  {
			
						console.log(res.data)
						
						}
						// FIN DU CATCH ERREUR \\
				})
			
			
				// FIN DE LA LACTION API \\
				//===========================
				//===========================
			}
		}
	}



module.exports = buttonroles;
